tests/lib/rules/jsx-child-element-spacing.js

```diff
@@ -0,0 +1,178 @@
+'use strict';
+
+const rule = require('../../../lib/rules/jsx-child-element-spacing');
+const RuleTester = require('eslint').RuleTester;
+const parserOptions = {
+  sourceType: 'module',
+  ecmaFeatures: {
+    jsx: true
+  }
+};
+
+const ERROR_MESSAGE = [{message: 'Ambiguous spacing between child elements.'}];
+
+const ruleTester = new RuleTester({parserOptions});
+ruleTester.run('jsx-child-element-spacing', rule, {
+  valid: [{
+    code: `
+      <App>
+        foo
+      </App>
+    `
+  }, {
+    code: `
+      <App>
+        <a>bar</a>
+      </App>
+    `
+  }, {
+    code: `
+      <App>
+        <a>
+          <b>nested</b>
+        </a>
+      </App>
+    `
+  }, {
+    code: `
+      <App>
+        foo
+        bar
+      </App>
+    `
+  }, {
+    code: `
+      <App>
+        <a>foo</a>
+        <a>bar</a>
+      </App>
+    `
+  }, {
+    code: `
+      <App>
+        foo<a>bar</a>baz
+      </App>
+    `
+  }, {
+    code: `
+      <App>
+        foo
+        {' '}
+        <a>bar</a>
+        {' '}
+        baz
+      </App>
+    `
+  }, {
+    code: `
+      <App>
+        foo
+        {' '}<a>bar</a>{' '}
+        baz
+      </App>
+    `
+  }, {
+    code: `
+      <App>
+        foo{' '}
+        <a>bar</a>
+        {' '}baz
+      </App>
+    `
+  }, {
+    code: `
+      <App>
+        foo{/*
+        */}<a>bar</a>{/*
+        */}baz
+      </App>
+    `
+  }, {
+    code: `
+      <App>
+        Please take a look at <a href="https://js.org">this link</a>.
+      </App>
+    `
+  }, {
+    code: `
+      <App>
+        Please take a look at
+        {' '}
+        <a href="https://js.org">this link</a>.
+      </App>
+    `
+  }, {
+    code: `
+      <App>
+        <p>A</p>
+        <p>B</p>
+      </App>
+    `
+  }, {
+    code: `
+      <App>
+        <p>A</p><p>B</p>
+      </App>
+    `
+  }, {
+    code: `
+      <App>
+        A
+        B
+      </App>
+    `
+  }],
+
+  invalid: [{
+    code: `
+      <App>
+        foo
+        <a>bar</a>
+      </App>
+    `,
+    errors: ERROR_MESSAGE
+  }, {
+    code: `
+      <App>
+        <a>bar</a>
+        baz
+      </App>
+    `,
+    errors: ERROR_MESSAGE
+  }, {
+    code: `
+      <App>
+        {' '}<a>bar</a>
+        baz
+      </App>
+    `,
+    errors: ERROR_MESSAGE
+  }, {
+    code: `
+      <App>
+        Please take a look at
+        <a href="https://js.org">this link</a>.
+      </App>
+    `,
+    errors: ERROR_MESSAGE
+  }, {
+    code: `
+      <App>
+        Here is
+        <a href="https://js.org">a link</a> and here is
+        <a href="https://js.org">another</a>
+      </App>
+    `,
+    errors: ERROR_MESSAGE
+  }, {
+    code: `
+      <App>
+        <a>
+          <b>nested1</b>
+          <b>nested2</b>
```

Comment:
why is this an error? are you going to keep a list of inline-by-default elements? What about when the display is changed via css?

Reply:
It would be an error because it would render as *nested1nested2* which is unlikely to be the developers intent

We certainly could just use the complete list of inline elements here: https://developer.mozilla.org/en-US/docs/Web/HTML/Inline_elements

Certainly it would not be feasible to capture changes to display by CSS. I wouldn't be overly concerned about it because...

- If the developer uses CSS to make a block element inline, then the rule will not catch this. However without this rule they are equally likely not to catch this. I would also expect that, since the developer is already modifying the default layout, they are likely to be visually inspecting the results manually to look for inconsistencies
- If the developer uses CSS to make an inline element into a block, then the rule may error unnecessarily. However the error can be silenced by adding either an explicit space or a comment form, as above.

Reply:
Unless there's already a package that keeps the list of inline elements up to date, i'd be hesitant to hardcode it in this package.

Reply:
A fair concern - here's one that I found: https://www.npmjs.com/package/block-elements (and it's less popular cousin https://www.npmjs.com/package/inline-elements)

Worth noting that React, when tackling a similar problem, opted to just hardcode a list of relevant tags

https://github.com/facebook/react/blob/17aa4d4682aa5c6fdbd3fb0d65e2d69769fe2d61/packages/react-dom/src/client/validateDOMNesting.js#L29

Another alternative would be not to do any special-casing on tag type and treat any adjacent tags as problematic, though I suspect that would get annoying pretty fast with adjacent divs.

Another concern - custom React classes for which it would be infeasible to detect the display type statically. This leads me to believe the rule should only report cases where we know the tag is inline - I think this covers a lot of common cases (like <a> tags in text)

Reply:
The real ideal would be if we could use React's list directly.

Reply:
Unfortunately, they are not the same use case

```
Note: this does not catch all invalid nesting, nor does it try to (as it's
  // not clear what practical benefit doing so provides); instead, we warn only
  // for cases where the parser will give a parse tree differing from what React
  // intended. For example, <b><div></div></b> is invalid but we don't warn
  // because it still parses correctly; we do warn for other cases like nested
  // <p> tags where the beginning of the second element implicitly closes the
  // first, causing a confusing mess.
```